Added Coulomb PP method and introduced a short-range cutoff

pwessels-uhh opened the pull request:
Added PP method to the Coulomb potential and introduced a short-range cutoff rs to deal with the divergence of the Coulomb potential for r -> 0. By default, rs is zero so that the cutoff is not in use. If electrons and ions are simulated explicitly using Coulomb PP without a short-range cutoff, unphysically fast electrons may emerge.

This pull request includes the absorbing boundary conditions as well as the expanded simulation box pull request so that all can be merged at once if deemed useful.

I am working on an example for the documentation close to the experiment in [T. Kroker et al., Nat. Commun. **12**, 596 (2021)] for a simulation of a plasma where the electron and ion trajectories are calculated explicitly to demonstrate all new features in this pull-request.

lucianogsilvestri commented:
Thanks a lot for the new feature. We have some concerns regarding the PP method for pure Coulomb potentials.  We think this is a good feature to add to our code, but if not used correctly it can cause problems. As you mentioned in your PR unphysical scenarios emerge when the cut-off is set to zero. So we would like to ask you if you are willing to work with us on modifying the documentation so that future users will use the code appropriately.  We noticed that you are preparing a notebook example and we greatly appreciate it. On a slightly different note. Sarkas has the ability to simulate open boundary conditions with pure and screened Coulomb potentials using Fast Multipole Method. This feature has been commented out as it was in the testing phase. It will soon be added  in the default version, we are waiting for the paper to be published. However, if you need to simulate electron-ion systems with open boundary conditions we can help you in running Sarkas with FMM. As a matter of fact the expanding plasma gif in the homepage of the docs was run with FMM.

pwessels-uhh commented:
Thanks a lot for the feedback! I agree with everything you said. I have added warning outputs if either the pure Coulomb PP method or the short-range cut-off are in use and also added a note to the documentation. Moreover, I integrated the ultrafast electron cooling example.

I also have added a dump for the potential energy per species and added a feature to include a monochromatic energy distribution where all particles have a fixed velocity but with a random unit vector distribution. I've also made a short note to the documentation. All these features are used and demonstrated in the UEC example.

On your side note: I will check out the FMM feature, looks like something we are looking for. And good luck with publishing the paper!

lucianogsilvestri commented:
Thank you very much for your work on improving Sarkas.
